### PR TITLE
Leaderboard

### LINKED ISSUE #851: create leaderboard service
Create new/update methods in in `server/src/services/leaderboard.ts` (and API endpoints in `server/src/routes/leaderboard.ts`) that will return global leaderboard/scoreboard data as well as personal player stat data

### Specifications
- create a shared (`shared/types.ts`) data structure to describe the leaderboard data
- if necessary, modify `PlayerStatItem` in `shared/types` as well
- create `LeaderboardService` method(s) to query the database with the typeorm API (https://typeorm.io/) and manipulate the data into the correct structure
- should return data for both bot and non-bot games (5 human players)
  - example of querying for games with/without bots can be found in the `GameService`
- create express GET endpoint in `routes/` so that data can be retrieved by the client
- limit response to 50/100 players
  - stretch goal: implement pagination

*If requests are too heavy, a solution may be to create a leaderboard table (entity) in the database that gets updated when a game finishes

#### Global Leaderboard Data Example

```ts
withBots: {
  [
    {
      rank: number,
      username: string,
      totalScore: number,
      victories: number,
      defeats: number,
    },
  ]
}

withoutBots: {
  [
    {
      rank: number,
      username: string,
      totalScore: number,
      victories: number,
      defeats: number,
    },
  ]
}
```

_use this issue for discussion/planning_

### LINKED ISSUE #852: create leaderboard page view
Create a new page in `client/src/views` that will display a tabular leaderboard/scoreboard as well as a player's personal stats if they are signed in (in a separate tab-style view)

#### Specifications
- use bootstrap-vue components such as `b-table`, etc. https://bootstrap-vue.org/docs
- use consistent styling (examples of pages with tables can be found in admin dashboard)
- consider responsiveness for different screen sizes in the page layout
- use data retrieved from the server side API #851
  - create a `LeaderboardAPI` in `client/src/api` to handle the requests with AJAX for the vue component
  - does not need to be 'real time', fetching the data once will be enough
  - should be able to choose between seeing the data for bot and non-bot games (5 human players)
- global leaderboard should be viewable without authentication at the route `/leaderboard/global`, `/leaderboard` should redirect here
- personal stats should be at `/leaderboard/stats` and should only be accessible by authenticated users
  - navigation tabs: https://bootstrap-vue.org/docs/components/nav#tab-style
  - router setup: https://github.com/virtualcommons/port-of-mars/blob/main/client/src/router.ts
  - old player stat component: https://github.com/virtualcommons/port-of-mars/blob/main/client/src/components/leaderboard/PlayerStatItem.vue

_use this issue for discussion/planning_

### PR DESCRIPTION
**Pull request by sgfost**
Adds global high scores and personal game history pages

resolves #852 
resolves #851

### CODE REVIEW

**Comment by sgfost** [comment]
Looking this over again, it seems like commit history got a bit mangled (probably by github desktop) with a merge creating duplicate commits, easily fixable but ought to be done before merging this

**Comment by Tkawamura02** [comment]
UPDATED TASKS FOR THIS ISSUE
- [ ] 1. fix any corner cases / bugs in the query logic that are causing the current leaderboard issues with the null points
- [ ] 2. Ensure that the only games ending in success count in the points total for a given player
- [ ] 3. Embed a short summary leaderboard view in the landing page as a contrasting section (after Gameplay?). This would only include the top 10 players and could be parameterizable Prop for the component
- [ ] 4. Minor styling improvements as needed

**Comment by saachibm** [comment]
> UPDATED TASKS FOR THIS ISSUE
> 
> * [ ]  1. fix any corner cases / bugs in the query logic that are causing the current leaderboard issues with the null points
> * [ ]  2. Ensure that the only games ending in success count in the points total for a given player
> * [ ]  3. Embed a short summary leaderboard view in the landing page as a contrasting section (after Gameplay?). This would only include the top 10 players and could be parameterizable Prop for the component
> * [ ]  4. Minor styling improvements as needed

Tiffany and I worked on task 2 and are continuing to figure out how to fix null points/corner cases (4/5/23)

**Comment by sgfost** [comment]
Just need to clean up the player stats/past games page for this to be ready to merge

**Review by alee — CHANGES_REQUESTED** [review]
Overall everything looks great but see the comments inline. Also I think there may be a bug in the player stats code (see screenshot) - the code in LeaderboardService looks OK though so I'm not exactly sure where it may be arising from.

image

File: server/src/routes/leaderboard.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { Router } from "express";
+import { getServices } from "@port-of-mars/server/services";
+import { User } from "@port-of-mars/server/entity";
+
+export const leaderboardRouter = Router();
+
+leaderboardRouter.get("/", async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (sgfost) [reply server/src/routes/leaderboard.ts]: Ended up resolving this alongside the other renaming, `stats` is now the base url/service name/category for directory naming purposes

File: client/src/views/Home.vue [hunk 2]
```diff
@@ -77,14 +86,32 @@
...
+            <h1 class="section-title mb-3">Community</h1>
+            <p class="text mb-3">
+              Whether you are looking to discuss the game or find a team, connect with other players
+              by joining our
+              <a :href="constants.DISCORD_URL">community Discord</a> or by joining a room in the
+              <b-link :to="lobby">game lobby</b-link> and using the built-in chat.
+            </p>
+            <p class="text mb-3">
+              Keep track of your performance with your
+              <b-link :to="playerStats">personal stats page</b-link> and visit the
```
[the remark this answers is not in the document]
Comment (sgfost) [reply client/src/views/Home.vue]: -> #857

File: client/src/views/PlayerStats.vue [hunk 3]
```diff
@@ -0,0 +1,41 @@
+<template>
+  <b-container fluid class="h-100 w-100 m-0 p-0 backdrop">
+    <div class="h-100 w-100 p-3 overflow-auto">
+      <b-row class="h-100 w-100 m-0">
+        <b-col cols="12" class="mh-100 p-2">
+          <PlayerStat v-for="item in playerStatItems" :key="item.time" :playerStatItem="item">
```
[the remark this answers is not in the document]
Comment (sgfost) [reply client/src/views/PlayerStats.vue]: -> #858

**Comment by sgfost** [comment]
Should be good to go @alee, anything not addressed in the last couple of commits are in new issues

**Review by alee — APPROVED** [review]
LGTM thanks all!

**Comment by alee** [comment]
@all-contributors
please add @sgfost for mentoring, code, doc, and design
please add @Tkawamura02 for code, doc, and design
please add @saachibm for code, doc, and design

**Pull request merged**